numpy/_core/src/umath/loops_trigonometric.dispatch.cpp

```diff
@@ -0,0 +1,298 @@
+    const vec_f32 invf2 = hn::Set(f32, -0x1.000000p-01f);
+    const vec_f32 invf0 = hn::Set(f32, 0x1.000000p+00f);
+
+    vec_f32 r = hn::MulAdd(invf8, x2, invf6);
+    r = hn::MulAdd(r, x2, invf4);
+    r = hn::MulAdd(r, x2, invf2);
+    r = hn::MulAdd(r, x2, invf0);
+    return r;
+}
+/*
+ * Approximate sine algorithm for x \in [-PI/4, PI/4]
+ * Maximum ULP across all 32-bit floats = 0.647
+ * Polynomial approximation based on unpublished work by T. Myklebust
+ */
+NPY_FINLINE vec_f32
+simd_sine_poly_f32(vec_f32& x, vec_f32& x2)
+{
+    const vec_f32 invf9 = hn::Set(f32, 0x1.7d3bbcp-19f);
+    const vec_f32 invf7 = hn::Set(f32, -0x1.a06bbap-13f);
+    const vec_f32 invf5 = hn::Set(f32, 0x1.11119ap-07f);
+    const vec_f32 invf3 = hn::Set(f32, -0x1.555556p-03f);
+
+    vec_f32 r = hn::MulAdd(invf9, x2, invf7);
+    r = hn::MulAdd(r, x2, invf5);
+    r = hn::MulAdd(r, x2, invf3);
+    r = hn::MulAdd(r, x2, hn::Zero(f32));
+    r = hn::MulAdd(r, x, x);
+    return r;
+}
+
+NPY_FINLINE vec_f32
+GatherIndexN(const float* src, npy_intp ssrc, npy_intp len)
+{
+    float temp[hn::Lanes(f32)] = { 0.0f };
+    for (auto ii = 0; ii < std::min(len, (npy_intp)hn::Lanes(f32)); ++ii) {
+        temp[ii] = src[ii * ssrc];
+    }
+    return hn::LoadU(f32, temp);
+}
+
+NPY_FINLINE void
+ScatterIndexN(vec_f32 vec, float* dst, npy_intp sdst, npy_intp len)
+{
+    float temp[hn::Lanes(f32)];
+    hn::StoreU(vec, f32, temp);
+    for (auto ii = 0; ii < std::min(len, (npy_intp)hn::Lanes(f32)); ++ii) {
+        dst[ii * sdst] = temp[ii];
+    }
+}
+
+static void SIMD_MSVC_NOINLINE
+simd_sincos_f32(const float *src, npy_intp ssrc, float *dst, npy_intp sdst,
+                npy_intp len, SIMD_TRIG_OP trig_op)
+{
+    // Load up frequently used constants
+    const vec_f32 zerosf = hn::Zero(f32);
+    const vec_s32 ones  = hn::Set(s32, 1);
+    const vec_s32 twos  = hn::Set(s32, 2);
+    const vec_f32 two_over_pi = hn::Set(f32, 0x1.45f306p-1f);
+    const vec_f32 codyw_pio2_highf = hn::Set(f32, -0x1.921fb0p+00f);
+    const vec_f32 codyw_pio2_medf = hn::Set(f32, -0x1.5110b4p-22f);
+    const vec_f32 codyw_pio2_lowf = hn::Set(f32, -0x1.846988p-48f);
+    const vec_f32 rint_cvt_magic = hn::Set(f32, 0x1.800000p+23f);
+    // Cody-Waite's range
+    float max_codi = 117435.992f;
+    if (trig_op == SIMD_COMPUTE_COS) {
+        max_codi = 71476.0625f;
+    }
+    const vec_f32 max_cody = hn::Set(f32, max_codi);
+
+    const int lanes = hn::Lanes(f32);
+    //npy_intp load_index[lanes/2];
+    //for (auto i = 0; i < lanes; ++i) {
+    //    load_index[i] = i * ssrc;
+    //}
+    //vec_s32 vec_lindex = hn::LoadU(s32, load_index);
+    //npy_intp store_index[lanes/2];
+    //for (auto i = 0; i < lanes; ++i) {
+    //    store_index[i] = i * sdst;
+    //}
+    //vec_s32 vec_sindex = hn::LoadU(s32, store_index);
+
+    for (; len > 0; len -= lanes, src += ssrc*lanes, dst += sdst*lanes) {
+        vec_f32 x_in;
+        if (ssrc == 1) {
```

Comment:
LoadN should only be used in the tail of a loop. It is probably worthwhile replicating the body of the loop (e.g. by moving it to a helper function), with most iterations using a normal load, and only the last iteration using LoadN.